### DISCUSSION TITLE
Noob-question: Typescript + Splitted Store + Middleware

### DISCUSSION THREAD

**Discussion by YeonV** [pmndrs/zustand#911] [Q&A] [answered]
Hello :)

THANK you soo much for Zustand! I love it!
Using it whereever I can :)
Now I would love to go Typescript.
Basic example for Typescript + Splitted Store + Middleware.

Goal:
![image](https://user-images.githubusercontent.com/28861537/163571708-a0854b4b-8024-4e8f-9ef8-f7948ec9c658.png)

I added 2 approaches for the `useStore`.
With each Im failing at some point:

```ts
// storeBears.tsx
import produce from 'immer'

export const storeBears = { 
  animals: { 
    bears: 1
  }
}

export const storeBearsActions = (set:any) => ({ 
  increase: (by: number):void => set(produce((state:any) => { state.animals.bears =  state.animals.bears + by }), false, "bears/increase"),
})
```

```ts
// storeUI.tsx

export const storeUI = { 
  darkMode: false,
}

export const storeUIActions = (set:any) => ({
  setDarkMode: (dark: boolean):void => set((state:any) => ({ darkMode: dark }), false, "ui/darkmode"),    
})
```

Approach 1:
```ts
// useStore.ts

import create from 'zustand';
import { devtools } from 'zustand/middleware';
import { combine } from 'zustand/middleware';
import { storeBears, storeBearsActions, } from './storeBears';
import { storeUI, storeUIActions } from './storeUI';

export const useStore = create(
  devtools(
    combine(
      {
        ...storeBears,
        ...storeUI
      },
      storeUIActions
      // >>>>cannot add storeBearsAction here<<<<
    )
  )
);
```

Approach 2:
```ts
// useStore.ts

import create from 'zustand';
import { devtools } from 'zustand/middleware';
import { combine } from 'zustand/middleware';
import { storeBears, storeBearsActions, } from './storeBears';
import { storeUI, storeUIActions } from './storeUI';

// >>>>cannot add middlewares here<<<<
export const useStore = create(((set) => ({  
  ...storeBearsActions(set),
  ...storeBears,
  
  ...storeUIActions(set),
  ...storeUI,
})))
```

**Comment by dai-shi** [comment]
Hi, have you see this page? https://github.com/pmndrs/zustand/wiki/Splitting-the-store-into-separate-slices

BTW, I'm not exactly sure why you can't add `devtools` in the second approach.

**Reply by YeonV** [reply]
> Hi, have you see this page? https://github.com/pmndrs/zustand/wiki/Splitting-the-store-into-separate-slices
> 
> BTW, I'm not exactly sure why you can't add `devtools` in the second approach.

Thanks for answering me <3

most likely because im a noob 🙈 

![image](https://user-images.githubusercontent.com/28861537/163573034-6ffdd940-4696-4e89-b61c-4e609f2b7baa.png)
![image](https://user-images.githubusercontent.com/28861537/163573054-3413d3f0-c6f6-47b2-a351-504bc2d4a059.png)

**Reply by YeonV** [reply]
Yes I saw that example, but I would love to have the "type inherit" benefit, without the need for an interface/type-definition

**Reply by dai-shi** [reply]
Oh, I missed your issue is TypeScript.

**Reply by dai-shi** [reply]
With v4, it will be improved. v3 requires to type store manually, and you need to learn from `tests/middlewateTypes.test.tsx`.

**Reply by YeonV** [reply]
I would like to create a working base first, before refactoring my apps:
https://github.com/YeonV/LedFx-Frontend-v2/blob/main/src/utils/apiStore.jsx
please dont laugh too much 🙈 its a mess and im a noob :)
![image](https://user-images.githubusercontent.com/28861537/163573884-81f19b54-7c97-41f6-8c7e-333f930c4e5f.png)
649 lines of noobness


Also I used Zustand for my [Vitron](https://github.com/YeonV/Vitron#vitron)-Project including [create-vitron](https://github.com/YeonV/create-vitron)

**Reply by YeonV** [reply]
> With v4, it will be improved. v3 requires to type store manually, and you need to learn from `tests/middlewateTypes.test.tsx`.

without middleware it works perfectly already <3
See my very first screenshot. Completely without manually typing 😎

**Reply by dai-shi** [reply]
Yes, with middleware, we need manual typing. (sorry, the main branch is already for v4, which is not yet released.)
https://github.com/pmndrs/zustand/blob/v3.7.2/tests/middlewareTypes.test.tsx#L126-L153

**Reply by YeonV** [reply]
@dai-shi hey broo,

i saw your v4-rc0 release <3

would one of the 2 approaches now work? :heart_eyes: 

Thx so much master!

**Reply by dai-shi** [reply]
Hope so. Please have a deep look at `./docs/*.md` and `./tests/middlewareTypes.test.tsx`.

**Comment by YeonV** [comment, marked as answer]
Ok got it working like this:

```js
import produce from 'immer'

export const storeBears = (set:any) => ({ 
  animals: { 
    bears: 1
  },
  increase: (by: number):void => set(produce((state:any) => { state.animals.bears =  state.animals.bears + by }), false, "bears/increase"),
})
```

```js
// import produce from 'immer'

export const storeUI = (set:any) => ({
  darkMode: false,
  setDarkMode: (dark: boolean):void => set((state:any) => ({ darkMode: dark }), false, "ui/darkmode"),    
})
```

```js
import create from 'zustand';
import { devtools } from 'zustand/middleware';
import { combine } from 'zustand/middleware';
import { storeBears } from './storeBears';
import { storeUI } from './storeUI';

export const useStore = create(
  devtools(
    combine(
      {
        hackedBy: 'Blade',
      },
      (set:any)=> ({
        ui: storeUI(set),
        example: storeBears(set)
      })
    )
  )
);
```
output:
![image](https://user-images.githubusercontent.com/28861537/164119293-adf927da-de9a-4551-a881-ba09d1076230.png)
